### PR TITLE
[rllib] Basic infrastructure for off-policy estimation (IS, WIS)

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Adds IS and WIS estimators, and an interface to easily add more (i.e. model-based ones) in the future.

This also patches a bunch of algorithms to return their current action probabilities, including adding a soft-Q option for DQN.

## Related issue number

Closes https://github.com/ray-project/ray/issues/3827
Closes https://github.com/ray-project/ray/issues/3942

### CODE REVIEW

**Comment by ericl** [comment]
FYI @joneswong @tengteng

File: python/ray/rllib/offline/shuffled_input.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def next(self):
+        if self.n <= 1:
+            return self.child.next()
+        if len(self.buffer) < self.n:
+            logger.info("Filling shuffle buffer to {} batches".format(self.n))
+            while len(self.buffer) < self.n:
+                self.buffer.append(self.child.next())
+            logger.info("Shuffle buffer filled")
+        i = random.randint(0, len(self.buffer) - 1)
+        self.buffer[i] = self.child.next()
```
Comment (richardliaw) [inline python/ray/rllib/offline/shuffled_input.py]: does this replace something?

Comment (ericl) [reply python/ray/rllib/offline/shuffled_input.py]: What do you mean, it's a random assign.

Comment (richardliaw) [reply python/ray/rllib/offline/shuffled_input.py]: Oh I guess I don't understand what you're doing here. Are you trying to do reservoir sampling? - https://en.wikipedia.org/wiki/Reservoir_sampling

**Comment by ericl** [comment]
No, I'm just adding a window buffer. There is no reservoir guarantee.

File: python/ray/rllib/offline/shuffled_input.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/offline/shuffled_input.py]: No, I'm just adding a window buffer. There is no reservoir guarantee.

File: python/ray/rllib/agents/marwil/marwil.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
     # Whether to rollout "complete_episodes" or "truncate_episodes"
     "batch_mode": "complete_episodes",
     # Read data from historic data and evaluate by a sampler
```
Comment (richardliaw) [inline python/ray/rllib/agents/marwil/marwil.py]: comment outdated?

**Review by richardliaw — APPROVED** [review]

**Comment by tengteng** [comment]
Great job! Thanks for implementing this. Is there any further plans for doubly robust estimator?

**Comment by ericl** [comment]
Thanks @tengteng . I filed https://github.com/ray-project/ray/issues/4018 for now, but probably won't be able to get to it for some time. Feel free to give it a shot if you need it earlier though.

**Pull request merged**